From e27aa746b9134fc14772c71617972b4cda3cd9c0 Mon Sep 17 00:00:00 2001
From: gengzonglin <gengzonglin@eswincomputing.com>
Date: Thu, 27 Mar 2025 10:13:12 +0800
Subject: [PATCH 108/109] WIN2030-17169:fix(bmc2.8):Fix the handling of escape
 char in web-server

Changelogs:
1.Fix the handling of escape char in web-server.

Signed-off-by: gengzonglin <gengzonglin@eswincomputing.com>
Change-Id: I55a056578f752c395a2f083c65cfd8b7e7d13f0d
---
 Core/Src/web-server.c | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/Core/Src/web-server.c b/Core/Src/web-server.c
index e8730f4..0a111f5 100644
--- a/Core/Src/web-server.c
+++ b/Core/Src/web-server.c
@@ -10821,6 +10821,33 @@ int set_rtcinfo(RTCInfo rtcInfo){
 	return 0;
 }
 
+char* url_decode(char* src)
+{
+	char* ori_str = src;
+    char* parse_str = strdup(src);
+    char* p = parse_str;
+    while (*src) {
+        if (*src == '%') {
+            if (src[1] && src[2]) {
+                /*Converts the two hexadecimal after the % to char.*/ 
+                sscanf(src + 1, "%2x", (unsigned int*)p++);
+                /*Skip %xx.*/
+                src += 3;
+                continue;
+            }
+        }
+        if (*src == '+') {
+            *p++ = ' ';
+            src++;
+        } else {
+            *p++ = *src++;
+        }
+    }
+    *p = '\0';
+	free(ori_str);
+    return parse_str;
+}
+
 //0,working,1,stopped
 int get_soc_status()
 {
@@ -11358,12 +11385,14 @@ http_server_netconn_serve(struct netconn *conn)
 					kv_pair *current = params.head;
 					while (current) {
 						if(strcmp(current->key,"username")==0){
+							current->value = url_decode(current->value);
 							username= current->value;
 							if(password!=NULL){//two param is found
 								break;
 							}
 						}
 						if(strcmp(current->key,"password")==0){
+							current->value = url_decode(current->value);
 							password= current->value;
 							if(username!=NULL){//two param is found
 								break;
@@ -11421,8 +11450,10 @@ http_server_netconn_serve(struct netconn *conn)
 					kv_pair *current = params.head;
 					while (current) {
 						if(strcmp(current->key,"password")==0){
+							current->value = url_decode(current->value);
 							password= current->value;
 						}else if(strcmp(current->key,"username")==0){
+							current->value = url_decode(current->value);
 							username= current->value;
 						}
 						current = current->next;
-- 
2.25.1

